### PR TITLE
changes to enable/disable metrics emission from user programs

### PR DESCRIPTION
**Pull request by gsps1**
Enabling/Disabling metrics emission from user programs
JIRA : https://issues.cask.co/browse/CDAP-12290

### CODE REVIEW

**Comment by gsps1** [comment]
build running : https://builds.cask.co/browse/CDAP-DUT6042-1

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 1]
```diff
@@ -206,6 +208,15 @@ protected AbstractContext(Program program, ProgramOptions programOptions, CConfi
     }
   }
 
+  protected MetricsCollectionService getMetricsService(CConfiguration cConf,
+                                                       MetricsCollectionService metricsCollectionService) {
+    if (cConf.getBoolean(Constants.Metrics.EMIT_PRGOGRAM_CONTAINER_METRICS,
+                         Constants.Metrics.DEFAULT_EMIT_PRGOGRAM_CONTAINER_METRICS)) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Don't define default in Constants. Always have it in `cdap-default.xml`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 2]
```diff
@@ -206,6 +208,15 @@ protected AbstractContext(Program program, ProgramOptions programOptions, CConfi
     }
   }
 
+  protected MetricsCollectionService getMetricsService(CConfiguration cConf,
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: why protected?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 3]
```diff
@@ -160,6 +161,7 @@ protected AbstractContext(Program program, ProgramOptions programOptions, CConfi
     this.programRunId = program.getId().run(ProgramRunners.getRunId(programOptions));
     this.discoveryServiceClient = discoveryServiceClient;
     this.owners = createOwners(program.getId());
+    metricsService = getMetricsService(cConf, metricsService);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Not a very good practice to modify a parameter. Maybe simply do:

```java
this.programMetrics = createProgramMetrics(
  programRunId, 
  cConf.getBoolean(...) ? metricsService : new NoOpMetricsCollectionService(), 
  metricsTags);
```

File: cdap-common/src/main/resources/cdap-default.xml [hunk 4]
```diff
@@ -2050,6 +2050,15 @@
     </description>
   </property>
 
+  <property>
+    <name>user.programs.emit.metrics</name>
```
Comment (chtyim) [inline cdap-common/src/main/resources/cdap-default.xml]: With the current naming convention, configurations that affect user program should be started with `app.program.`. Also, for boolean config, it should be ended with `.enabled`. Maybe call it `app.program.metrics.enabled`

Comment (anwar6953) [reply cdap-common/src/main/resources/cdap-default.xml]: if so, remove 'disabled' from the description.

**Comment by gsps1** [comment]
@chtyim thanks for the review, i have addressed the comments, please take a look again.

**Comment by chtyim** [comment]
LGTM

**Pull request merged**